src/server/dflycmd.cc

```diff
@@ -54,177 +55,209 @@ void DflyCmd::Run(CmdArgList args, ConnectionContext* cntx) {
-      return rb->SendError(kInvalidSyncId);
-    }
+  // assuming here that shard id and thread id is the same thing.
+  listener_->Migrate(cntx->owner(), shard_set->pool()->at(threadid));
 
-    // Before setting up a transaction that initiates a full sync,
-    // we kick-off non-shard replication that must start earlier than shards.
-    shard_set->pool()->AwaitFiberOnAll([&](unsigned index, auto*) {
-      auto* shard = EngineShard::tlocal();
-      if (!shard) {
-        StartReplInThread(index, syncid);
-      }
-    });
-
-    cntx->transaction->Schedule();
+  // response is an array
+  rb->StartArray(2);
+  rb->SendSimpleString("FULL");
+  rb->SendSimpleString(eof_token);
+}
 
-    AggregateStatus status;
-    // kick off the snapshotting simultaneously in all shards.
-    // we do it via data sockets but reply here
-    // via control socket that orchestrates the flow from the replica side.
-    cntx->transaction->Execute(
-        [&](Transaction* t, EngineShard* shard) {
-          status = FullSyncInShard(syncid, t, shard);
-          return OpStatus::OK;
-        },
-        true);
+void DflyCmd::Sync(CmdArgList args, ConnectionContext* cntx) {
+  RedisReplyBuilder* rb = static_cast<RedisReplyBuilder*>(cntx->reply_builder());
 
-    // TODO: after the full sync we should continue streaming data from non-shard threads as well
-    // because those threads provide locking/transaction information needed to provide atomicity
-    // guarantees on replica.
+  auto sync_data = GetSyncInfoOrRespond(ArgS(args, 2), rb);
+  if (!sync_data)
+    return;
+  auto [sync_id, sync_info] = *sync_data;
 
-    if (*status == OpStatus::OK)
-      return rb->SendOk();
+  VLOG(0) << "Starting full sync on " << sync_id;
+  {
+    std::unique_lock lk(mu_);
 
-    // TODO: to stop the replication in shards that started doing this.
-    // We can do it without the transaction via shard_set.
+    if (sync_info->state != ReplicaState::PREPARATION) {
+      return rb->SendError("?");
+    }
 
-    return rb->SendError("replication error");
+    // Initialize barrier. -1 for io_thread, +1 for last kick.
+    unsigned flows = sync_info->thread_map.size();
+    sync_info->snapshot_block.emplace(flows, &sync_info->fiber_flag_counter);
+    sync_info->fiber_flag_counter.store(flows, std::memory_order_relaxed);
   }
-  rb->SendError(kSyntaxErr);
-}
 
-uint32_t DflyCmd::AllocateSyncSession() {
-  unique_lock lk(mu_);
-  auto [it, inserted] = sync_info_.emplace(next_sync_id_, new SyncInfo);
-  CHECK(inserted);
+  // Before setting up a transaction that initiates a full sync,
+  // we kick-off non-shard replication that must start earlier than shards.
+  // TODO: do in parallel with full sync fibers when transmitting header data.
+  shard_set->pool()->AwaitFiberOnAll([this, sync_id = sync_id](unsigned index, auto*) {
+    auto* shard = EngineShard::tlocal();
+    if (!shard) {
+      StartReplInThread(index, sync_id);
+    }
+  });
```

Comment:
See TODO. 
If this flow transmits header values (lua scripts), then it should run in parallel?